### PR TITLE
New Pre-Process

### PR DESCRIPTION
**Pull request by martinyang0416**
2024.03.25 Updates and Comments
1. 3D structure is controlled by ignore_3D in Mordred feature 
2. Added GetChiralTag as One Hot Encoding 
3. Add clean_feature function to drop mostly missing columns and fill the rest of the missing values with column mean.

### CODE REVIEW

File: src/preprocess.py [hunk 1]
```diff
@@ -129,25 +144,67 @@ def extract_attributes_from_df(df):
...
...fr_nitro', 'fr_nitro_arom', 'fr_nitro_arom_nonortho', 'fr_nitroso', 'fr_oxazole', 'fr_oxime', 'fr_para_hydroxylation', 'fr_phenol', 'fr_phenol_noOrthoHbond', 'fr_phos_acid', 'fr_phos_ester', 'fr_piperdine', 'fr_piperzine', 'fr_priamide', 'fr_prisulfonamd', 'fr_pyridine', 'fr_quatN', 'fr_sulfide', 'fr_sulfonamd', 'fr_sulfone', 'fr_term_acetylene', 'fr_tetrazole', 'fr_thiazole', 'fr_thiocyan', 'fr_thiophene', 'fr_unbrch_alkane', 'fr_urea', 'BalabanJ', 'BertzCT', 'LabuteASA', 'PEOE_VSA1', 'PEOE_VSA2', 'PEOE_VSA3', 'PEOE_VSA4', 'PEOE_VSA5', 'PEOE_VSA6', 'PEOE_VSA7', 'PEOE_VSA8', 'PEOE_VSA9', 'PEOE_VSA10', 'PEOE_VSA11', 'PEOE_VSA12', 'PEOE_VSA13', 'SMR_VSA1', 'SMR_VSA2', 'SMR_VSA3', 'SMR_VSA4', 'SMR_VSA5', 'SMR_VSA6', 'SMR_VSA7', 'SMR_VSA8', 'SMR_VSA9', 'SlogP_VSA1', 'SlogP_VSA2', 'SlogP_VSA3', 'SlogP_VSA4', 'SlogP_VSA5', 'SlogP_VSA6', 'SlogP_VSA7', 'SlogP_VSA8', 'SlogP_VSA9', 'SlogP_VSA10', 'SlogP_VSA11', 'EState_VSA1', 'EState_VSA2', 'EState_VSA3', 'EState_VSA4', 'EState_VSA5', 'EState_VSA6', 'EState_VSA7', 'EState_VSA8', 'EState_VSA9', 'EState_VSA10', 'VSA_EState1', 'VSA_EState2', 'VSA_EState3', 'VSA_EState4', 'VSA_EState5', 'VSA_EState6', 'VSA_EState7', 'VSA_EState8', 'VSA_EState9']
```
Comment (davidanugraha) [inline src/preprocess.py:147]: Too messy, refactor variable names to top of the file

File: src/preprocess.py [hunk 2]
```diff
@@ -101,11 +112,15 @@ def smile_to_graph(smile):
...
         # Create features for each atom based on its symbol, degree, number of Hs, and implicit valence
-        feature = (one_hot_encoding(atom.GetSymbol(), ELEMENT_LIST) + one_hot_encoding(atom.GetDegree(), DEGREE_LIST) +
-                   one_hot_encoding(atom.GetTotalNumHs(), NUMBER_OF_H_LIST) + one_hot_encoding(atom.GetImplicitValence(), IMPLICIT_VALENCE_LIST))
-
+        # feature = (one_hot_encoding(atom.GetSymbol(), ELEMENT_LIST) + one_hot_encoding(atom.GetDegree(), DEGREE_LIST) +
+        #            one_hot_encoding(atom.GetTotalNumHs(), NUMBER_OF_H_LIST) + one_hot_encoding(atom.GetImplicitValence(), IMPLICIT_VALENCE_LIST))
+        feature = [atom.GetAtomicNum(), atom.GetMass(),atom.GetDegree(),atom.GetTotalNumHs(), atom.GetImplicitValence(),atom.GetExplicitValence(),
+               atom.GetFormalCharge(),atom.GetIsAromatic(), atom.GetIsotope(), atom.GetNoImplicit(), atom.GetNumExplicitHs(),atom.GetNumImplicitHs(),
+               atom.GetNumRadicalElectrons(),atom.GetTotalDegree(), atom.GetTotalValence(), atom.InvertChirality(), atom.IsInRing()
+               ]
```
Comment (davidanugraha) [inline src/preprocess.py:120]: Missing GetChiralTag

File: src/preprocess.py (hunk 1 above)
Comment (davidanugraha) [inline src/preprocess.py:147]: Too messy, refactor variable names to top of the file

Comment (davidanugraha) [reply src/preprocess.py:147]: Also have you tested this? Are you sure these values are numerically stable. In addition, are 3D Features included in here as well?

**Review by davidanugraha — CHANGES_REQUESTED** [review]

File: src/preprocess.py (hunk 2 above)
Comment (derrickkkkk) [reply src/preprocess.py:120]: GetChiralTag is not Numerical

Comment (davidanugraha) [reply src/preprocess.py:120]: Yes, have to make it Categorical as I mentioned

File: src/preprocess.py [hunk 3]
```diff
@@ -13,11 +13,15 @@
...
 ELEMENT_LIST = ['C', 'N', 'O', 'S', 'F', 'Si', 'P', 'Cl', 'Br', 'Mg', 'Na','Ca', 'Fe', 'As', 'Al', 'I', 'B', 'V', 'K', 'Tl', 'Yb','Sb', 'Sn', 
                 'Ag', 'Pd', 'Co', 'Se', 'Ti', 'Zn', 'H','Li', 'Ge', 'Cu', 'Au', 'Ni', 'Cd', 'In', 'Mn', 'Zr','Cr', 'Pt', 'Hg', 'Pb', 'Other']
+
 DEGREE_LIST = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
 NUMBER_OF_H_LIST = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
 IMPLICIT_VALENCE_LIST = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+ChiralTags = ['CHI_UNSPECIFIED', 'CHI_TETRAHEDRAL_CW', 'CHI_TETRAHEDRAL_CCW', 'CHI_OTHER']
```
Comment (davidanugraha) [inline src/preprocess.py:23]: Are you sure these are all the ChiralTags? I remember it's more than this, please refer to documentation of Rdkit please

Comment (davidanugraha) [inline src/preprocess.py:23]: [ChiralType](https://www.rdkit.org/docs/cppapi/classRDKit_1_1Atom.html#a8a82ae947ebbbc48f5ec5128f5c3e724) {
  [CHI_UNSPECIFIED](https://www.rdkit.org/docs/cppapi/classRDKit_1_1Atom.html#a8a82ae947ebbbc48f5ec5128f5c3e724a964d8c1bfb4dd140e8c56538df248139) = 0 , [CHI_TETRAHEDRAL_CW](https://www.rdkit.org/docs/cppapi/classRDKit_1_1Atom.html#a8a82ae947ebbbc48f5ec5128f5c3e724a8c09352c68561c82ac07f27f75911fdd) , [CHI_TETRAHEDRAL_CCW](https://www.rdkit.org/docs/cppapi/classRDKit_1_1Atom.html#a8a82ae947ebbbc48f5ec5128f5c3e724aa95768d8f526ebcd9f01410b73a1e6fd) , [CHI_OTHER](https://www.rdkit.org/docs/cppapi/classRDKit_1_1Atom.html#a8a82ae947ebbbc48f5ec5128f5c3e724a29e23be90b99e1255fe2ddb4be817157) ,
  [CHI_TETRAHEDRAL](https://www.rdkit.org/docs/cppapi/classRDKit_1_1Atom.html#a8a82ae947ebbbc48f5ec5128f5c3e724a6e4dc33ab4e36387e517e7e6cf3ba598) , [CHI_ALLENE](https://www.rdkit.org/docs/cppapi/classRDKit_1_1Atom.html#a8a82ae947ebbbc48f5ec5128f5c3e724a74d2d5b724ce5022c7b7f2bbae28f497) , [CHI_SQUAREPLANAR](https://www.rdkit.org/docs/cppapi/classRDKit_1_1Atom.html#a8a82ae947ebbbc48f5ec5128f5c3e724af64e94d024b146c6147ef561ca228e7e) , [CHI_TRIGONALBIPYRAMIDAL](https://www.rdkit.org/docs/cppapi/classRDKit_1_1Atom.html#a8a82ae947ebbbc48f5ec5128f5c3e724ab3840bb1ccb4b294a85ea327b34d94ea) ,
  [CHI_OCTAHEDRAL](https://www.rdkit.org/docs/cppapi/classRDKit_1_1Atom.html#a8a82ae947ebbbc48f5ec5128f5c3e724a9f32ced132f353561e5dcd81582fe609)
}

Comment (davidanugraha) [reply src/preprocess.py:23]: Also please fix naming to all capital

File: src/preprocess.py [hunk 4]
```diff
@@ -107,6 +111,8 @@ def one_hot_encoding(x, l):
     # Return one-hot encoded list
     return list(map(lambda s: x == s, l))
 
+ChiralTags = ['CHI_UNSPECIFIED', 'CHI_TETRAHEDRAL_CW', 'CHI_TETRAHEDRAL_CCW', 'CHI_OTHER']
```
Comment (davidanugraha) [inline src/preprocess.py:114]: Duplicated code?

File: src/preprocess.py [hunk 5]
```diff
@@ -117,7 +123,7 @@ def smile_to_graph(smile):
         feature = [atom.GetAtomicNum(), atom.GetMass(),atom.GetDegree(),atom.GetTotalNumHs(), atom.GetImplicitValence(),atom.GetExplicitValence(),
                atom.GetFormalCharge(),atom.GetIsAromatic(), atom.GetIsotope(), atom.GetNoImplicit(), atom.GetNumExplicitHs(),atom.GetNumImplicitHs(),
                atom.GetNumRadicalElectrons(),atom.GetTotalDegree(), atom.GetTotalValence(), atom.InvertChirality(), atom.IsInRing()
-               ]
+               ] + one_hot_encoding(atom.GetChiralTag(), ChiralTags)
```
Comment (davidanugraha) [inline src/preprocess.py:126]: Please use pandas

File: src/preprocess.py [hunk 6]
```diff
@@ -13,11 +13,15 @@
...
...fr_nitro', 'fr_nitro_arom', 'fr_nitro_arom_nonortho', 'fr_nitroso', 'fr_oxazole', 'fr_oxime', 'fr_para_hydroxylation', 'fr_phenol', 'fr_phenol_noOrthoHbond', 'fr_phos_acid', 'fr_phos_ester', 'fr_piperdine', 'fr_piperzine', 'fr_priamide', 'fr_prisulfonamd', 'fr_pyridine', 'fr_quatN', 'fr_sulfide', 'fr_sulfonamd', 'fr_sulfone', 'fr_term_acetylene', 'fr_tetrazole', 'fr_thiazole', 'fr_thiocyan', 'fr_thiophene', 'fr_unbrch_alkane', 'fr_urea', 'BalabanJ', 'BertzCT', 'LabuteASA', 'PEOE_VSA1', 'PEOE_VSA2', 'PEOE_VSA3', 'PEOE_VSA4', 'PEOE_VSA5', 'PEOE_VSA6', 'PEOE_VSA7', 'PEOE_VSA8', 'PEOE_VSA9', 'PEOE_VSA10', 'PEOE_VSA11', 'PEOE_VSA12', 'PEOE_VSA13', 'SMR_VSA1', 'SMR_VSA2', 'SMR_VSA3', 'SMR_VSA4', 'SMR_VSA5', 'SMR_VSA6', 'SMR_VSA7', 'SMR_VSA8', 'SMR_VSA9', 'SlogP_VSA1', 'SlogP_VSA2', 'SlogP_VSA3', 'SlogP_VSA4', 'SlogP_VSA5', 'SlogP_VSA6', 'SlogP_VSA7', 'SlogP_VSA8', 'SlogP_VSA9', 'SlogP_VSA10', 'SlogP_VSA11', 'EState_VSA1', 'EState_VSA2', 'EState_VSA3', 'EState_VSA4', 'EState_VSA5', 'EState_VSA6', 'EState_VSA7', 'EState_VSA8', 'EState_VSA9', 'EState_VSA10', 'VSA_EState1', 'VSA_EState2', 'VSA_EState3', 'VSA_EState4', 'VSA_EState5', 'VSA_EState6', 'VSA_EState7', 'VSA_EState8', 'VSA_EState9']
```
Comment (davidanugraha) [inline src/preprocess.py:16]: Same, make it capital. Also, I don't see any HOMO, LUMO attributes for 3D, where is it?

File: src/preprocess.py [hunk 7]
```diff
@@ -171,14 +184,34 @@ def calculate_mordred_features(smile_string):
...
     values_list = []
     for value in descriptor_values.values():
-        if isinstance(value, (int, float, np.number)) and not np.isnan(value):
-            values_list.append(value)
-        else:
-            values_list.append(0) 
+        # if isinstance(value, (int, float, np.number)) and not np.isnan(value):
+        values_list.append(value)
+        # else:
+        #     values_list.append(0)
```
Comment (davidanugraha) [inline src/preprocess.py:190]: Commented codes should be cleaned and removed

File: src/preprocess.py [hunk 8]
```diff
@@ -144,22 +150,29 @@ def extract_attributes_from_df(df):
...
+        print(f"Error: Invalid SMILES '{smiles}'.")
+        return None
     try:
         property_value = getattr(Descriptors, property_name)(molecule)
-        return property_value
+        # Ensure the property value is numerical
+        if isinstance(property_value, (int, float)):
+            return property_value
+        else:
+            print(f"Warning: Property '{property_name}' did not return a numerical value.")
```
Comment (davidanugraha) [inline src/preprocess.py:164]: Use logging

**Review by davidanugraha — CHANGES_REQUESTED** [review]
Please see the comments and fix them.

**Pull request merged**